### PR TITLE
Set Deimos lamps parset from header

### PR DESCRIPTION
**Pull request by debora-pe**
This is to allow PypeIt to read from the header the lamps that were used when the arcs where taken and set those as the `lamps` parset for the wavelength calibration.

The main changes are in `keck_deimos.py`.

### CODE REVIEW

**Comment by debora-pe** [comment]
Tests pass.

Dev Suite
```
--- PYPEIT DEVELOPMENT SUITE PASSED 128/128 TESTS (Masters ignored) ---
Testing Started at 2021-09-22T18:19:15.051665
Testing Completed at 2021-09-22T23:20:16.274794
Total Time: 5:01:01.223129
```
Unit tests
`====================== 289 passed in 1011.90s (0:16:51) =======================`

File: pypeit/pypeit.py [hunk 1]
```diff
@@ -113,7 +113,7 @@ def __init__(self, pypeit_file, verbosity=2, overwrite=True, reuse_masters=False
         if config_specific_file is not None:
             msgs.info(
                 'Setting configuration-specific parameters using {0}'.format(os.path.split(config_specific_file)[1]))
-        spectrograph_cfg_lines = self.spectrograph.config_specific_par(config_specific_file).to_config()
+        spectrograph_cfg_lines = self.spectrograph.config_specific_par(config_specific_file, pypeit_file=pypeit_file).to_config()
```
Comment (debora-pe) [inline pypeit/pypeit.py]: @kbwestfall @jhennawi Thank you for your feedback today.
I like the idea of using the `fitstbl` but I'm still not convinced that it will not needed to be created twice.
Once, before the `config_specific_par` (before line 116). And in this case, `self.par` (which is required for `fitstbl`) could be created using the `default_pypeit_par`. Then  `spectrograph_cfg_lines` are generated and `self.par` must be generated again as it's done in line 120, because here makes the merge with the user-proved parsets (giving the preference to the users parsets). And since `self.par` changes, I am guessing that the `fitstbl` should be generated again. Am I making sense?

Comment (tepickering) [reply pypeit/pypeit.py]: fwiw, the ability to parse lamps out of FITS headers would also be very useful for MMTO Blue Channel. HeNeAr is by far the most commonly used and is what is currently hard-coded. however, ThAr is used with some frequency and FeNe and HgCd are available and used for specific configurations. would be desirable to only include line lists for lamps that are actually turned on for a specific image. so i appreciate the groundwork being laid out to make it possible.

File: pypeit/pypeit.py [hunk 2]
```diff
@@ -268,6 +268,8 @@ def calib_all(self):
             # Find all the frames in this calibration group
             in_grp = self.fitstbl.find_calib_group(i)
             grp_frames = frame_indx[in_grp]
+            if self.par['calibrations']['wavelengths']['use_header_lamps']:
+                self.par['calibrations']['wavelengths']['lamps'] = self.spectrograph.get_lamps(self.fitstbl[in_grp])
```
Comment (kbwestfall) [inline pypeit/pypeit.py]: Can we simplify the `get_lamps` function by only passing the arc rows of the metadata table here?  I.e., something like:
```
if self.par['calibrations']['wavelengths']['use_header_lamps']:
    is_arc = self.fitstbl.find_frames('arc')
    self.par['calibrations']['wavelengths']['lamps'] \
            = self.spectrograph.get_lamps(self.fitstbl[in_grp & is_arc])
```

File: pypeit/spectrographs/keck_deimos.py [hunk 3]
```diff
@@ -716,6 +718,28 @@ def bpm(self, filename, det, shape=None, msbias=None):
 
         return bpm_img
 
+    def get_lamps(self, fitstbl):
```
Comment (kbwestfall) [inline pypeit/spectrographs/keck_deimos.py:721]: Is there anything in this function that's specifically tied to DEIMOS?  I.e., apart from enforcing that this approach only be supported for DEIMOS, is there a reason we can't make this an attribute of the `Spectrograph` base class?

File: pypeit/core/qa.py [hunk 4]
```diff
@@ -45,7 +45,7 @@ def set_qa_filename(root, method, det=None, slit=None, prefix=None, out_dir=None
         outfile = 'QA/PNGs/Slit_Profile_{:s}_'.format(root)
     elif method == 'arc_fit_qa':
 #        outfile = 'QA/PNGs/Arc_1dfit_{:s}_S{:04d}.png'.format(root, slit)
-        outfile = 'PNGs/Arc_1dfit_{:s}_S{:04d}.png'.format(root, slit)
+        outfile = 'PNGs/Arc_1dfit_{:s}_S{:04d}.png'.format(root, slit) if slit is not None else 'PNGs/Arc_1dfit_{:s}.pdf'.format(root)
```
Comment (kbwestfall) [inline pypeit/core/qa.py]: Just to check: Does this affect the output QA html files that are written?

**Review by kbwestfall — COMMENTED** [review]
Thanks for the changes @deborape .  I had a couple more thoughts for your consideration.

File: pypeit/core/qa.py (hunk 4 above)
Comment (debora-pe) [reply pypeit/core/qa.py]: Thanks @kbwestfall  for looking into these changes, and raising this point. I haven't fully tested this yet. I was going to do it today.
But just to give you some context, the reason I have done that is that I am hoping to share these plots with KOA whenever we share our reduced ADAP data and I though they needed to be more easily shareable.

File: pypeit/pypeit.py (hunk 2 above)
Comment (debora-pe) [reply pypeit/pypeit.py]: Thanks, I'll try that!

File: pypeit/spectrographs/keck_deimos.py (hunk 3 above)
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py:721]: Actually, It is somewhat DEIMOS specific, because it depends on how the lamps are called in the header of the specific instrument and how is translated in "PypeIt language". 
As you can see line 739 is doing this conversion, i.e., taking the name of the lamp from the header and adding a "I", which works for all the DEIMOS lamps, but it's a bit more complicated for some of the other instruments.

Comment (kbwestfall) [reply pypeit/spectrographs/keck_deimos.py:721]: Thanks for checking!  That makes perfect sense.

File: pypeit/pypeit.py (hunk 2 above)
Comment (debora-pe) [reply pypeit/pypeit.py]: Done!

File: pypeit/core/qa.py (hunk 4 above)
Comment (debora-pe) [reply pypeit/core/qa.py]: Thanks for this comment! I decided to revert these changes. As you guessed this affected the QA html, i.e., the wavelength fits were not saved in the html file. I decided that for KOA probably it will be easier to create html files for only the wavelength calibration (these are the only QA that I think it's important to share), but I realize I will need some help with that.

File: pypeit/spectrographs/keck_deimos.py [hunk 5]
```diff
@@ -716,6 +718,27 @@ def bpm(self, filename, det, shape=None, msbias=None):
...
+
+        """
+        # read and save the arc lamp in each frame
+        lamp_list = np.array([])
+        for col in fitstbl:
+                lamp_list = np.append(lamp_list, col['lampstat01'].split(' '))
+        # make it in Pypeit format
+        lamp_list = [lamp + 'I' for lamp in np.unique(lamp_list)]
+
+        return lamp_list
```
Comment (kbwestfall) [inline pypeit/spectrographs/keck_deimos.py]: Sorry to keep nit-picking, but you can make this whole function a one-liner:

```
return [f'{x}I' for x in np.unique(np.concatenate([l.split() for l in fitstbl['lampstat01']]))]
```

**Review by kbwestfall — APPROVED** [review]
Thanks for this!  I've got one last nit-pick, but happy to approve now.

File: pypeit/spectrographs/keck_deimos.py (hunk 5 above)
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py]: Thanks! I made this change.

File: pypeit/core/wavecal/autoid.py [hunk 6]
```diff
@@ -158,14 +158,16 @@ def arc_fit_qa(waveFit, outfile=None, ids_only=False, title=None,
 
     # Stats
     wave_soln_fit = waveFit.pypeitfit.eval(waveFit.pixel_fit/waveFit.xnorm)#, 'legendre',minx=fit['fmin'], maxx=fit['fmax'])
-    rms = np.sqrt(np.sum((waveFit.wave_fit[gpm]-wave_soln_fit[gpm])**2)/len(waveFit.pixel_fit[gpm])) # Ang
-    dwv_pix = np.median(np.abs(waveFit.wave_soln-np.roll(waveFit.wave_soln,1)))
-    ax_fit.text(0.1*len(arc_spec), 0.90*ymin+(ymax-ymin),r'$\Delta\lambda$={:.3f}$\AA$ (per pix)'.format(dwv_pix), size='small')
-    ax_fit.text(0.1*len(arc_spec), 0.80*ymin+(ymax-ymin),'RMS={:.3f} (pixels)'.format(rms/dwv_pix), size='small')
+    # DP: Adjusted rms here because sometimes does not exactly matches the value that we record in
```
Comment (profxj) [inline pypeit/core/wavecal/autoid.py]: Good catch.  I think you can remove those lines.

File: pypeit/spectrographs/keck_deimos.py [hunk 7]
```diff
@@ -293,6 +293,8 @@ def config_specific_par(self, scifile, inp_par=None):
             par['calibrations']['wavelengths']['method'] = 'full_template'
             par['calibrations']['wavelengths']['reid_arxiv'] = 'keck_deimos_900ZD.fits'
             par['calibrations']['wavelengths']['lamps'] += ['CdI', 'ZnI', 'HgI']
+        # Arc lamps list from header
+        par['calibrations']['wavelengths']['use_header_lamps'] = True
```
Comment (profxj) [inline pypeit/spectrographs/keck_deimos.py]: Can you add a brief comment or two somewhere
on this being the default for DEIMOS, 
e.g. https://pypeit.readthedocs.io/en/release/deimos.html ?

**Review by profxj — CHANGES_REQUESTED** [review]
Please add the doc request I made and then
feel free to merge (after running DEIMOS tests).

Thanks!

File: pypeit/core/wavecal/autoid.py (hunk 6 above)
Comment (debora-pe) [reply pypeit/core/wavecal/autoid.py]: Done!

File: pypeit/spectrographs/keck_deimos.py (hunk 7 above)
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py]: Done!

**Comment by debora-pe** [comment]
Tests passes.

Dev Suite 
```
--- PYPEIT DEVELOPMENT SUITE PASSED 128/128 TESTS (Masters ignored) ---
Testing Started at 2021-10-07T11:26:09.727999
Testing Completed at 2021-10-07T16:08:28.818411
Total Time: 4:42:19.090412
```
Unit tests.
`================ 302 passed in 1244.17s (0:20:44) ==========================`

**Review by profxj — APPROVED** [review]
Thumbs up from me!

File: pypeit/pypeit.py [hunk 8]
```diff
@@ -279,6 +279,10 @@ def calib_all(self, run=True):
             # Find all the frames in this calibration group
             in_grp = self.fitstbl.find_calib_group(i)
             grp_frames = frame_indx[in_grp]
+            if self.par['calibrations']['wavelengths']['use_header_lamps']:
```
Comment (jhennawi) [inline pypeit/pypeit.py]: Can these lines instead go in Calibrations rather than here. You have the fitstbl in that routine, and there you already isolate which files are the ArcFrame, so it seems more appropriate to do this there. 

As a secondary issue, this would also have the advantage that you don't need to modify the lamps at all in the parset, i.e. you can just use

lamps = self.par['calibrations']['wavelengths']['lamps'] if not self.par['calibrations']['wavelengths']['use_header_lamps'] else self.spectrograph.get_lamps(self.fitstbl[in_grp & is_arc])

In other words, it seems that everyone is in agreement that we should move toward parsets being immutable, in which case we should pass lamps in as a keyword argument to the wavecalib routines.

File: pypeit/pypeit.py [hunk 9]
```diff
@@ -357,9 +361,12 @@ def reduce_all(self):
 
         # Iterate over each calibration group and reduce the standards
         for i in range(self.fitstbl.n_calib_groups):
-
             # Find all the frames in this calibration group
             in_grp = self.fitstbl.find_calib_group(i)
+            if self.par['calibrations']['wavelengths']['use_header_lamps']:
+                is_arc = self.fitstbl.find_frames('arc')
+                self.par['calibrations']['wavelengths']['lamps'] = \
+                    self.spectrograph.get_lamps(self.fitstbl[in_grp & is_arc])
```
Comment (jhennawi) [inline pypeit/pypeit.py]: Same comment as above

**Review by jhennawi — CHANGES_REQUESTED** [review]
I think you can move the main control flow change to Calibrations rather than in pypeit. This would be cleaner and you wouldn't need to search for the arc files twice.

**Comment by debora-pe** [comment]
OK, once again all the tests passed.

Dev Suite 
```
--- PYPEIT DEVELOPMENT SUITE PASSED 129/129 TESTS (Masters ignored) ---
Testing Started at 2021-10-13T18:41:33.705792
Testing Completed at 2021-10-14T00:18:57.429077
Total Time: 5:37:23.723285
```

Unit tests
`==================== 302 passed in 1276.03s (0:21:16) =====================`

File: CHANGES.rst [hunk 10]
```diff
@@ -2,6 +2,8 @@
 --------
 
 - Introduces pypeit_parse_calib_id script
+- Pypeit can now read the list of arc lamps from the header and use it for
+  wavelength calibration.
```
Comment (kbwestfall) [inline CHANGES.rst]: Almost not worth mentioning, but I would specify that this is currently for DEIMOS only.

**Review by kbwestfall — APPROVED** [review]

File: CHANGES.rst (hunk 10 above)
Comment (debora-pe) [reply CHANGES.rst]: Oh yeah! Thanks!!!

File: pypeit/core/gui/identify.py [hunk 11]
```diff
@@ -195,6 +195,8 @@ def initialise(cls, arccen, slits, slit=0, par=None, wv_calib_all=None,
         ----------
         arccen : ndarray
             Arc spectrum
+        lamps : :obj:`list`
```
Comment (jhennawi) [inline pypeit/core/gui/identify.py:198]: Can you provide an example of what the input looks like here for users that may want to launch this from a notebook or IPython shell.

File: pypeit/core/wavecal/autoid.py [hunk 12]
```diff
@@ -861,6 +858,8 @@ def full_template(spec, par, ok_mask, det, binspectral, nsnippet=2, debug_xcorr=
     Args:
         spec: ndarray (nspec, nslit)
           Spectra to be calibrated
+        lamps : :obj:`list`
```
Comment (jhennawi) [inline pypeit/core/wavecal/autoid.py:861]: Same comment here as above, please provide an example of lamps in the docs.

File: pypeit/core/wavecal/autoid.py [hunk 13]
```diff
@@ -1063,6 +1062,8 @@ class ArchiveReid:
         desired.
     spectrograph : :class:`~pypeit.spectrographs.spectrograph.Spectrograph`
         Spectrograph instance
+    lamps : :obj:`list`
```
Comment (jhennawi) [inline pypeit/core/wavecal/autoid.py:1066]: Also here please.

File: pypeit/core/wavecal/autoid.py [hunk 14]
```diff
@@ -1336,6 +1337,8 @@ class HolyGrail:
     ----------
     spec : ndarray
         2D array of arcline spectra (nspec,nslit)
+    lamps : :obj:`list`
```
Comment (jhennawi) [inline pypeit/core/wavecal/autoid.py:1342]: Also here.

File: pypeit/core/wavecal/autoid.py [hunk 15]
```diff
@@ -1376,15 +1379,15 @@ class HolyGrail:
...
+    def __init__(self, spec, lamps, par=None, ok_mask=None, islinelist=False,
+                 outroot=None, debug=False, verbose=False,
                  binw=None, bind=None, nstore=1, use_unknowns=True, 
                  nonlinear_counts=None, spectrograph=None):
 
         # Set some default parameters
         self._spec = spec
         self._par = pypeitpar.WavelengthSolutionPar() if par is None else par
-        self._lines = self._par['lamps']
+        self._lines = lamps
```
Comment (jhennawi) [inline pypeit/core/wavecal/autoid.py]: Can we redefine the variable self._lines to be self._lamps to be more consistent.

File: pypeit/core/wavecal/waveio.py [hunk 16]
```diff
@@ -239,7 +239,7 @@ def load_line_lists(lines, unknown=False, skip=False, all=False, NIST=False,
 
     Parameters
     ----------
-    lamps : list
+    lines : list
```
Comment (jhennawi) [inline pypeit/core/wavecal/waveio.py]: It would be preferable to have lamps refer to a list of arc lamps (e.g. ['Th', 'Ar']), and lines to be a list of lines. I'm not actually sure what lines is supposed to contain in these codes, but I'm imagining it to be a set of wavelengths, whereas lamps is a set of lines. It would be easier to follow these codes if consistent terminology is used.

File: pypeit/core/wavecal/waveio.py [hunk 17]
```diff
@@ -253,6 +253,7 @@ def load_line_lists(lines, unknown=False, skip=False, all=False, NIST=False,
     line_list : Table
 
     """
+    msgs.info("Arc lamps used: {}".format(', '.join(lines)))
```
Comment (jhennawi) [inline pypeit/core/wavecal/waveio.py]: Based on this I think we want to call lines lamps.

File: pypeit/par/pypeitpar.py [hunk 18]
```diff
@@ -2358,7 +2358,9 @@ def __init__(self, reference=None, method=None, echelle=None, ech_fix_format=Non
         #options['lamps'] = WavelengthSolutionPar.valid_lamps()
         dtypes['lamps'] = list
         descr['lamps'] = 'Name of one or more ions used for the wavelength calibration.  Use ' \
-                         'None for no calibration.  ' # \
+                         '``None`` for no calibration. Choose ``use_header`` to use the list of lamps ' \
+                         'recoded in the header of the arc frames (this is currently ' \
```
Comment (jhennawi) [inline pypeit/par/pypeitpar.py]: recoded --> "recorded"

File: pypeit/scripts/identify.py [hunk 19]
```diff
@@ -94,12 +94,12 @@ def main(args):
 
         # Load the MasterFrame (if it exists and is desired).  Bad-pixel mask
         # set to any flagged pixel in MasterArc.
-        wavecal = BuildWaveCalib(msarc, slits, spec, par, binspectral=slits.binspec, det=args.det,
+        wavecal = BuildWaveCalib(msarc, slits, spec, par, lamplist, binspectral=slits.binspec, det=args.det,
```
Comment (jhennawi) [inline pypeit/scripts/identify.py]: Can we call this lamps to be consistent with the naming of other variables in other routines?

File: pypeit/scripts/identify.py [hunk 20]
```diff
@@ -94,12 +94,12 @@ def main(args):
...
                                  master_key=mkey, msbpm=msarc.select_flag())
         arccen, arc_maskslit = wavecal.extract_arcs(slitIDs=[args.slit])
 
         # Launch the identify window
-        arcfitter = Identify.initialise(arccen, slits, slit=int(args.slit), par=par,
+        arcfitter = Identify.initialise(arccen, lamplist, slits, slit=int(args.slit), par=par,
```
Comment (jhennawi) [inline pypeit/scripts/identify.py]: lamplist --> lamps

File: pypeit/spectrographs/keck_deimos.py [hunk 21]
```diff
@@ -278,7 +278,7 @@ def config_specific_par(self, scifile, inp_par=None):
         if self.get_meta_value(headarr, 'dispname') == '600ZD':
             par['calibrations']['wavelengths']['method'] = 'full_template'
             par['calibrations']['wavelengths']['reid_arxiv'] = 'keck_deimos_600ZD.fits'
```
Comment (jhennawi) [inline pypeit/spectrographs/keck_deimos.py:280]: Two things. Shouldn't we be setting par['calibrations']['wavelengths']['lamps'] = 'use_header' somewhere here? Also, I'm a bit lost as to how the lamps can depend on what is written in the header, but not the reid_arxiv? Won't you get garbage results if lamps and reid_arxiv are not matched?

**Review by jhennawi — APPROVED** [review]
Hi @deborape, 

Thanks for this. Everything looks very good. I have relatively minor comments on variable names and doc modifications that I'd appreciate you look at, but once that is done you can merge. There is a substantive comment regarding the DEIMOS reid_arxiv and lamps. If this is an issue, it need not be addressed in this PR, i.e. you can just respond here or in Slack and make an issue if it is something we need to address in the future.

File: pypeit/core/gui/identify.py (hunk 11 above)
Comment (debora-pe) [reply pypeit/core/gui/identify.py:198]: Done!

File: pypeit/core/wavecal/autoid.py (hunk 12 above)
Comment (debora-pe) [reply pypeit/core/wavecal/autoid.py:861]: Done

File: pypeit/core/wavecal/autoid.py (hunk 13 above)
Comment (debora-pe) [reply pypeit/core/wavecal/autoid.py:1066]: Done.

File: pypeit/core/wavecal/autoid.py (hunk 14 above)
Comment (debora-pe) [reply pypeit/core/wavecal/autoid.py:1342]: Done.

File: pypeit/core/wavecal/autoid.py (hunk 15 above)
Comment (debora-pe) [reply pypeit/core/wavecal/autoid.py]: Done.

File: pypeit/core/wavecal/waveio.py (hunk 16 above)
Comment (debora-pe) [reply pypeit/core/wavecal/waveio.py]: Changed "lines" into "lamps"

File: pypeit/core/wavecal/waveio.py (hunk 17 above)
Comment (debora-pe) [reply pypeit/core/wavecal/waveio.py]: Done

File: pypeit/par/pypeitpar.py (hunk 18 above)
Comment (debora-pe) [reply pypeit/par/pypeitpar.py]: Done

File: pypeit/scripts/identify.py (hunk 19 above)
Comment (debora-pe) [reply pypeit/scripts/identify.py]: Done

File: pypeit/scripts/identify.py (hunk 20 above)
Comment (debora-pe) [reply pypeit/scripts/identify.py]: Done

File: pypeit/spectrographs/keck_deimos.py (hunk 21 above)
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py:280]: Currently we don't have a different template for each combination of lamps. All templates (except for 830G) have been created using all the possible DEIMOS arc lamps. In general, this work well even if the users use less lamps.

**Comment by debora-pe** [comment]
@kbwestfall @profxj @jhennawi 
I just wanted to make sure I did not make any mistakes in the latest changes to address Joe's comments (I have changed the name of a few variables), so I re-run the tests. All passed.
```
--- PYPEIT DEVELOPMENT SUITE PASSED 129/129 TESTS (Masters ignored) ---
Testing Started at 2021-10-25T23:01:27.655244
Testing Completed at 2021-10-26T04:41:11.710232
Total Time: 5:39:44.054988
```
`========================= 302 passed in 1270.04s (0:21:10) =========================`

I'm merging now.
